b2/transferer.py

```diff
@@ -8,23 +8,104 @@
+from .exception import ChecksumMismatch, UnexpectedCloudBehaviour, TruncatedOutput
 from .progress import DoNothingProgressListener
+from .raw_api import SRC_LAST_MODIFIED_MILLIS
+
+# block size used when downloading file. If it is set to a high value, progress reporting will be jumpy, if it's too low, it impacts CPU
+BLOCK_SIZE = 4096
 
 
 class Transferer(object):
-    """ Manage downloads and uploads """
+    """ Handles complex actions around downloads and uploads to free raw_api from that responsibility """
 
     def __init__(self, session, account_info):
         self.session = session
         self.account_info = account_info
 
     def download_file_from_url(self, url, download_dest, progress_listener=None, range_=None):
+        """
+        :param url: url from which the file should be downloaded from
+        :param download_dest: where to put the file when it is downloaded
+        :param progress_listener: where to notify about progress downloading
+        :param range: 2-element tuple containing data of http Range header
+        """
         progress_listener = progress_listener or DoNothingProgressListener()
-        with progress_listener:
-            return self.session.download_file_from_url(
-                url,
-                DownloadDestProgressWrapper(download_dest, progress_listener),
-                url_factory=self.account_info.get_download_url,
-                range_=range_,
+        download_dest = DownloadDestProgressWrapper(download_dest, progress_listener)
+        with self.session.download_file_from_url(
+            url,
+            url_factory=self.account_info.get_download_url,
+            range_=range_,
+        ) as response:
+
+            info = response.headers
+
+            file_id = info['x-bz-file-id']
+            file_name = info['x-bz-file-name']
+            content_type = info['content-type']
+            content_length = int(info['content-length'])
+            content_sha1 = info['x-bz-content-sha1']
+            if range_ is not None:
+                if 'Content-Range' not in info:
+                    raise UnexpectedCloudBehaviour('Content-Range header was expected')
+            file_info = dict((k[10:], info[k]) for k in info if k.startswith('x-bz-info-'))
+
+            digest = hashlib.sha1()
+            bytes_read = 0
+
+            info_dict = self._download_response_to_info_dict(response)
+
+            mod_time_millis = int(
+                info_dict['fileInfo'].get(
+                    SRC_LAST_MODIFIED_MILLIS,
+                    info['x-bz-upload-timestamp'],
+                )
             )
+
+            with download_dest.make_file_context(
+                file_id,
+                file_name,
+                content_length,
+                content_type,
+                content_sha1,
+                file_info,
+                mod_time_millis,
+                range_=range_,
+            ) as file:
+                with progress_listener:
```

Comment:
Seems like the DownloadDestProgressWrapper should take care of closing the progress listener.

Reply:
The progress bar situation is severe, I'll make a separate PR for fixing this